src/NUnitFramework/framework/Constraints/UniqueItemsConstraint.cs

```diff
@@ -60,5 +69,123 @@ protected override bool Matches(IEnumerable actual)
+            if (memberType == null || IsHandledSpeciallyByNUnit(memberType))
+                return null;
+
+            string methodName = nameof(ItemsUnique);
+            bool makeGeneric = true;
+
+            // Special handling for ignore case with strings and chars
+            if (IgnoringCase)
+            {
+                if (memberType == typeof(string))
+                {
+                    methodName = nameof(StringsUniqueIgnoringCase);
+                    makeGeneric = false;
+                }
+                else
+                if (memberType == typeof(char))
+                {
+                    methodName = nameof(CharsUniqueIgnoringCase);
+                    makeGeneric = false;
+                }
+            }
+
+            MethodInfo method = GetType().GetMethod(methodName, BindingFlags.Static | BindingFlags.NonPublic);
+            if (method == null)
+                throw new InvalidOperationException("Internal error - Method not found: " + methodName);
+            
+            if (makeGeneric)
+                method = method.MakeGenericMethod(memberType);
+
+            return (bool)method.Invoke(null, new object[] { actual });
+        }
+
+        private static bool ItemsUnique<T>(IEnumerable<T> actual)
+        {
+            var hash = new HashSet<T>();
+
+            foreach (T item in actual)
+            {
+                if (hash.Contains(item))
+                    return false;
+
+                hash.Add(item);
+            }
+
+            return true;
+        }
+
+        private static bool StringsUniqueIgnoringCase(IEnumerable<string> actual)
+        {
+            var hash = new HashSet<string>();
+
+            foreach (string item in actual)
+            {
+                string s = item.ToLower();
+                if (hash.Contains(s))
+                    return false;
+
+                hash.Add(s);
+            }
+
+            return true;
+        }
+
+        private static bool CharsUniqueIgnoringCase(IEnumerable<char> actual)
+        {
+            var hash = new HashSet<char>();
+
+            foreach (char item in actual)
+            {
+                char ch = char.ToLower(item);
+                if (hash.Contains(ch))
+                    return false;
+
+                hash.Add(ch);
+            }
+
+            return true;
+        }
+
+        // Return true if NUnitEqualityHandler has special logic for Type
+        private static bool IsHandledSpeciallyByNUnit(Type type)
+        {
+            if (type == typeof(string)) return false; // even though it's IEnumerable
+
+            return type.IsArray
+                || type == typeof(IEnumerable) // Covers lists, collections, dictionaries as well
+                || type == typeof(System.IO.Stream)
+                || type == typeof(System.IO.DirectoryInfo)
+                || type.FullName == "System.Tuple"
+                || type.FullName == "System.ValueTuple";
+        }
```

Comment:
How do we ensure that this list doesn't get stale?
Also `type == typeof(IEnumerable)` doesn't cover collections etc, but `typeof(IEnumerable).IsAssignableFrom(type)` does.

Reply:
I don't think staleness is a big problem, since any change in the definition of NUnit equality would be a pretty big breaking change.

Regarding `type == typeof`` you are right. Not sure if I can use `IsAssignableFrom` in all our builds, but I'll give it a shot.

Reply:
We added ValueTuple and later Tuple and if this list had existed back then and Mikkel hadn't run across it, I'm imagining it being missed in review and indefinitely.

I think importing the reflection compatibility namespace will handle `IsAssignableFrom`.